File: 3. Behavioral Design Patterns/Visitor/Java/VisitorExample.java
// Visitor Pattern Example in Java
// Document structure with different element types processed by visitors

import java.util.*;

interface DocumentVisitor {
    void visit(Paragraph paragraph);
    void visit(Image image);
    void visit(Table table);
}

interface DocumentElement {
    void accept(DocumentVisitor visitor);
}

class Paragraph implements DocumentElement {
    private final String text;
    public Paragraph(String text) { this.text = text; }
    public String getText() { return text; }
    public void accept(DocumentVisitor visitor) { visitor.visit(this); }
}

class Image implements DocumentElement {
    private final String filename;
    public Image(String filename) { this.filename = filename; }
    public String getFilename() { return filename; }
    public void accept(DocumentVisitor visitor) { visitor.visit(this); }
}

class Table implements DocumentElement {
    private final int rows, columns;
    public Table(int rows, int columns) { this.rows = rows; this.columns = columns; }
    public int getRows() { return rows; }
    public int getColumns() { return columns; }
    public void accept(DocumentVisitor visitor) { visitor.visit(this); }
}

class WordCountVisitor implements DocumentVisitor {
    private int wordCount = 0;
    public int getWordCount() { return wordCount; }
    public void visit(Paragraph paragraph) { wordCount += paragraph.getText().split("\\s+").length; }
    public void visit(Image image) { /* images don't contribute to word count */ }
    public void visit(Table table) { wordCount += table.getRows() * table.getColumns(); /* assume 1 word per cell */ }
}

class PrintVisitor implements DocumentVisitor {
    public void visit(Paragraph paragraph) { System.out.println("Paragraph: " + paragraph.getText()); }
    public void visit(Image image) { System.out.println("Image: " + image.getFilename()); }
    public void visit(Table table) { System.out.println("Table: " + table.getRows() + "x" + table.getColumns()); }
}

public class VisitorExample {
    public static void main(String[] args) {
        List<DocumentElement> elements = Arrays.asList(
            new Paragraph("Hello world from paragraph one."),
            new Image("photo.jpg"),
            new Table(3, 4),
            new Paragraph("This is another paragraph.")
        );

        WordCountVisitor wordCounter = new WordCountVisitor();
        PrintVisitor printer = new PrintVisitor();

        for (DocumentElement element : elements) {
            element.accept(wordCounter);
            element.accept(printer);
        }

        System.out.println("Total word count: " + wordCounter.getWordCount());
    }
}

File: 3. Behavioral Design Patterns/Visitor/Kotlin/VisitorExample.kt
// Visitor Pattern Example in Kotlin
// Document structure with different element types processed by visitors

interface DocumentVisitor {
    fun visit(paragraph: Paragraph)
    fun visit(image: Image)
    fun visit(table: Table)
}

interface DocumentElement {
    fun accept(visitor: DocumentVisitor)
}

class Paragraph(val text: String) : DocumentElement {
    override fun accept(visitor: DocumentVisitor) = visitor.visit(this)
}

class Image(val filename: String) : DocumentElement {
    override fun accept(visitor: DocumentVisitor) = visitor.visit(this)
}

class Table(val rows: Int, val columns: Int) : DocumentElement {
    override fun accept(visitor: DocumentVisitor) = visitor.visit(this)
}

class WordCountVisitor : DocumentVisitor {
    var wordCount = 0
        private set
    override fun visit(paragraph: Paragraph) { wordCount += paragraph.text.split("\\s+".toRegex()).size }
    override fun visit(image: Image) { /* images don't contribute to word count */ }
    override fun visit(table: Table) { wordCount += table.rows * table.columns /* assume 1 word per cell */ }
}

class PrintVisitor : DocumentVisitor {
    override fun visit(paragraph: Paragraph) = println("Paragraph: ${paragraph.text}")
    override fun visit(image: Image) = println("Image: ${image.filename}")
    override fun visit(table: Table) = println("Table: ${table.rows}x${table.columns}")
}

fun main() {
    val elements = listOf(
        Paragraph("Hello world from paragraph one."),
        Image("photo.jpg"),
        Table(3, 4),
        Paragraph("This is another paragraph.")
    )

    val wordCounter = WordCountVisitor()
    val printer = PrintVisitor()

    for (element in elements) {
        element.accept(wordCounter)
        element.accept(printer)
    }

    println("Total word count: ${wordCounter.wordCount}")
}
